text/0152-telemetry-schemas.md

```diff
@@ -0,0 +1,1174 @@
+Each edge in the tree represents a schema transformation.
+
+Once we have the tree of schemas it is possible to then convert from a node in a
+tree to another node in a tree given that the nodes are connected via a path
+through edges.
+
+Further research is needed in this area if we find that we need this concept of
+parent schemas.
+
+### Current State in Schema
+
+The schema file in this OTEP describes changes from version to version. This is
+very useful for performing conversions between versions but it does not capture
+the current state of a particular version.
+
+We can add an ability to specify the full current schema of each particular
+version in the schema file (in addition to the "changes" section). This will
+open up a few interesting possibilities:
+
+- Automatic validation that emitted telemetry confirms to the declared schema.
+
+- OpenTelemetry semantic conventions document can be automatically generated
+  from this formal schema file. This will remove the need to have semantic
+  conventions codified and generated from yaml files in the specification.
+
+- Consumers of telemetry can use this information to interpret the received
+  telemetry.
+
+- It may be possible to auto-generate the "changes" section in schema files as a
+  delta between full states of versions.
+
+An example of what can be recorded as the "current state" is the encoding of the
+Body of the Log Records. It was described as a concept in
+[this earlier proposal](https://docs.google.com/document/d/1ZExye1lW43owwaxcbjOvl0P2qER-UaYd_MyxItc2h0k/edit#).
+The suggestion was to record the encoding of the Body field of Log Records,
+which can be also implemented as a new setting in the "current state" of the
+"logs" section in the schema file.
+
+### Other Transformation Types
+
+This OTEP introduces a limited set of transformations, while deliberately
+keeping the type of transformations to a minimum. It is easy to see that
+telemetry can evolve in other, more sophisticated ways over time (see e.g.
+[this list](https://github.com/open-telemetry/opentelemetry-specification/issues/1324)).
+
+More transformation types can be added in the future when there is a need to
+represent a particular change in the telemetry schema that is not possible with
+the current set of transformations. However, care should be taken to avoid
+introducing unnecessary or overly complicated transformation types, especially
+ones that cannot be applied on a local portion of telemetry data and requires a
+full state (e.g. transformations, such as aggregations, etc), since such
+transformations can place significant implementation burden on telemetry sources
+that wish to support the notion of telemetry schemas.
+
+## Alternates Considered
+
+### Freeze Schema
+
+Instead of introducing formal schemas, schema files and version we can require
+that the instrumentation once it is created never changes. Attributes are never
```

Comment:
As long as the instrumentation conforms to the semantic conventions before and after, that should be fine, shouldn't it?

I think rather than freezing the schema completely, we only need to ban changes that would create an entry in the YAML file defined in this OTEP. So renaming is not allowed, adding new attributes is. Changing the meaning of an attribute is also not allowed, but I think this can also not automatically handled with the schema YAML, although the version specified there could then be used in the backend for version-specific handling.

Reply:
Yes, certain changes would be fine with the approach you propose. I still don't think this alternate approach is sufficient. I don't believe it solves the problems we want to solve. In my opinion being able to rename attributes or metrics is a basic requirement. Given how much time we spend arguing about attribute names here in Otel semantic conventions and how much uncertainty is there that the name is right even after the attribute is accepted I believe we will inevitably run into a situation when name changes are wanted.

I saw schema troubles in the past. For example with metrics emitted by Otel Collector, we renamed metrics and broke Collector dashboards in a proprietary monitoring tool. When schemas are not explicitly declared and are not a visible concept they still change and things break, just less explicitly visibly so.

Again, we can argue that we should be careful and come up with semantic conventions that we think are right and then never change them. I think it is an attempt to achieve perfection that is very difficult to do and at the same time increases the barrier to introducing semantic conventions because the cost of making a mistake is very high. I think we should lower that cost a bit. It should still be expensive to change semantic conventions (we don't want to encourage that too much), but it should not be impossible.